src/pybind/mgr/telegraf/module.py

```diff
@@ -0,0 +1,315 @@
+
+    def init_module_config(self):
+        self.config['address'] = \
+            self.get_config("address", default=self.config_keys['address'])
+        self.config['interval'] = \
+            int(self.get_config("interval",
+                                default=self.config_keys['interval']))
+
+    def now(self):
+        return int(round(time.time() * 1000000000))
+
+    def send_to_telegraf(self):
+        url = urlparse(self.config['address'])
+
+        sock = BaseSocket(url)
+        self.log.debug('Sending data to Telegraf at %s', sock.address)
+        sock.connect()
+
+        now = self.now()
+
+        try:
+            measurements = list()
+            measurements += self.get_pool_stats()
+            measurements += self.get_daemon_stats()
+            measurements += self.get_cluster_stats()
+
+            for measurement in measurements:
+                self.log.debug(measurement)
+                line = Line(measurement['measurement'], measurement['value'],
+                            measurement['tags'], now)
+                self.log.debug(line.to_line_protocol())
+                sock.send(line.to_line_protocol())
+
+        except (socket.error, RuntimeError):
+            self.log.exception('Failed to send statistics to Telegraf:')
+        finally:
+            sock.close()
+
+    def shutdown(self):
+        self.log.info('Stopping Telegraf module')
+        self.run = False
+        self.event.set()
+
+    def handle_command(self, cmd):
+        if cmd['prefix'] == 'telegraf config-show':
+            return 0, json.dumps(self.config), ''
+        elif cmd['prefix'] == 'telegraf config-set':
+            key = cmd['key']
+            value = cmd['value']
+            if not value:
+                return -errno.EINVAL, '', 'Value should not be empty or None'
+
+            self.log.debug('Setting configuration option %s to %s', key, value)
+            self.set_config_option(key, value)
+            self.set_config(key, value)
+            return 0, 'Configuration option {0} updated'.format(key), ''
+        elif cmd['prefix'] == 'telegraf send':
+            self.send_to_telegraf()
+            return 0, 'Sending data to Telegraf', ''
+        if cmd['prefix'] == 'telegraf self-test':
+            return 0, '', 'Self-test OK'
+
+        return (-errno.EINVAL, '',
+                "Command not found '{0}'".format(cmd['prefix']))
+
+    def serve(self):
+        self.log.info('Starting Telegraf module')
+        self.init_module_config()
+        self.run = True
+
+        self.log.debug('Waiting 10 seconds before starting')
+        self.event.wait(10)
+
+        while self.run:
+            try:
+                start = self.now()
+                self.send_to_telegraf()
+                runtime = (self.now() - start) / 1000000
+                self.log.debug('Sending data to Telegraf took %d ms', runtime)
+            except:
+                self.log.exception('Failed to send statistics to Telegraf:')
```

Comment:
Since send_to_telegraf already has its own exception handling around the network IO, do we need this catchall handler?  Any exception getting to this point is probably something buggy/unexpected, so maybe it's better to just let it fall out and show up as a module error in `ceph status`